README.md
# erl-memcache

This library is a wrapper around poolboy and erlmc_conn. The first one is used to create a pool of connections, the latter 
library is included and creates a persistent connection to Memcached (binary protocol).

Features include:
 * automatically start a local Memcached instance
 * (named) pools and their management
 * statistics retrieval from Memcached


Example configuration:
```
[
	{memcache, 
		[{pools,
			[{yaddapool, [
                {size, 10},
                {max_overflow, 20},
                {port, 3333},
                {host, "localhost"},
                {opts, [{memory, 4}]},
                {start_server, true}
            ]}]
        }]
    }
]
```

Notes:
 * currently only the memory of a memcached instance (-m) can be set. 

DOC/CHANGELOG
Current version: 1.2.0
    NEW support for memcached options, currently limited to memory size
    NEW support for memcached options, currently limited to memory size

tag: 1.1.2
    NEW memcache:get_pools/0
    FIX Environment preconfiguration fixes
    NEW memcache:get_pools/0
    FIX properly register memcached instances
    FIX children startup order

tag: 1.1.1
    NEW set memcache pools via environment config
    NEW set memcache pools via environment config

tag: 1.1.0
    NEW memcache:stats/1

tag: 1.0.10
    NEW erlanglibs 2.7.3

tag: 1.0.9
    OPT Improved a testcase with a retry mechanism
    OPT memcached availability verification

tag: 1.0.8
    OPT timeout increased for one test
    FIX some extra time given to memcached to die after kill -9

tag: 1.0.7
    CHANGE shared pools functionality rolled back
    NEW Allowing pools sharing address
    FIX added error handling for race condition in closing sock

tag: 1.0.6

tag: 1.0.5
    FIX Another attempt to improve behaviour on CI
    FIX Attempt to fix socket reuse issue on jenkins
    OPT Added unit test for single pool restart

tag: 1.0.4
    FIX graceful poolboy shutdown

tag: 1.0.3
    DOC added overview.edoc
    OPT poolboy dependency updated

tag: 1.0.2
    FIX poolboy version verified in rebar.config

tag: 1.0.1
    FIX supervisor report when stopping a pool

tag: 1.0.0
    OPT Added forgotten test module
    OPT Error handling improved when a pool dies
    CHANGE emptied just before updating the newest version
    NEW First fully functional version
    FIX dialyzer passing
    NEW Initial version

tag: 0.1.0
    FIX configs
    NEW more tests
    NEW added tests
    FIX erlanglibs version
    CHANGE Added config and runner scripts.
    FIX dependencies poolboy and erlanglibs


test/memcache_eunit.erl
-module(memcache_eunit).

-include_lib("eunit/include/eunit.hrl").

%%================================================================================================
%% TESTS
%%================================================================================================

application_start_test_() ->
    ?_test(begin
            error_logger:tty(false),
            {ok, F} = elibs_application:start(memcache),
            ?assertMatch(P when is_pid(P), whereis(memcache_sup)),
            ?assertMatch(P when is_pid(P), whereis(memcache_pools_sup)),
            ?assertMatch(P when is_pid(P), whereis(memcache)),
            ok = F(),
            ?assertMatch(undefined, whereis(memcache_sup)),
            ?assertMatch(undefined, whereis(memcache_pools_sup)),
            ?assertMatch(undefined, whereis(memcache))
        end).

single_pool_operation_test_() ->
    {setup, fun setup/0, fun cleanup/1, fun (_) ->
                StartPoolRes=memcache:start_pool(testpool, "localhost", 3333, 10, 10, true),
                StartPoolSameName=memcache:start_pool(testpool, "localhost", 3334, 10, 10, true),
                PoolNotFoundOp=memcache:get(unknownpool, <<"key">>),
                EmptyGet=memcache:get(testpool, <<"key">>),
                EmptyGet2=memcache:get(testpool, key),
                Set=memcache:set(testpool, <<"key">>, <<"value">>),
                SetExpiring=memcache:set(testpool, <<"key3">>, <<"value3">>, 1),
                Readback=memcache:get(testpool, <<"key">>),
                Readback2=memcache:get(testpool, key),
                timer:sleep(1000),
                ReadbackExpired=memcache:get(testpool, key3),
                AnotherSet=memcache:set(testpool, anotherkey, <<"anothervalue">>),
                Delete=memcache:delete(testpool, anotherkey),
                DeleteNotFound=memcache:delete(testpool, asdasdasd),
                ReadbackDeleted=memcache:get(testpool, anotherkey),
                Flush=memcache:flush(testpool),
                ReadbackFlushed=memcache:get(testpool, <<"key">>),
                StopPool=memcache:stop_pool(testpool),
                StopPoolNotStarted=memcache:stop_pool(testpool2),
                [?_assertEqual(ok, StartPoolRes),
                 ?_assertEqual({error, poolname_in_use}, StartPoolSameName),
                 ?_assertEqual({error, pool_not_found}, PoolNotFoundOp),
                 ?_assertEqual({ok, <<>>}, EmptyGet),
                 ?_assertEqual({ok, <<>>}, EmptyGet2),
                 ?_assertEqual({ok, <<"value">>}, Set),
                 ?_assertEqual({ok, <<"value3">>}, SetExpiring),
                 ?_assertEqual({ok, <<"value">>}, Readback),
                 ?_assertEqual({ok, <<>>}, Readback2),
                 ?_assertEqual({ok, <<>>}, ReadbackExpired),
                 ?_assertEqual({ok, <<"anothervalue">>}, AnotherSet),
                 ?_assertEqual(ok, Delete),
                 ?_assertEqual(ok, DeleteNotFound),
                 ?_assertEqual({ok, <<>>}, ReadbackDeleted),
                 ?_assertEqual(ok, Flush),
                 ?_assertEqual({ok, <<>>}, ReadbackFlushed),
                 ?_assertEqual(ok, StopPool),
                 ?_assertEqual({error, pool_not_found}, StopPoolNotStarted)]
        end}.

pool_independence_test_() ->
    {setup, fun setup/0, fun cleanup/1, fun (_) ->
                StartPool1=memcache:start_pool(testpool, "localhost", 3333, 10, 10, true),
                StartPool2=memcache:start_pool(testpool2, "localhost", 3334, 10, 10, true),
                Set1=memcache:set(testpool, key, <<"value">>),
                Set2=memcache:set(testpool2, key, <<"value2">>),
                Readback1=memcache:get(testpool, key),
                Readback2=memcache:get(testpool2, key),
                Flush1=memcache:flush(testpool),
                ReadbackFlushed1=memcache:get(testpool, key),
                ReadbackFlushed2=memcache:get(testpool2, key),
                GetPools=memcache:get_pools(),
                StopPool1=memcache:stop_pool(testpool),
                StopPool2=memcache:stop_pool(testpool2),
                [?_assertEqual(ok, StartPool1),
                 ?_assertEqual(ok, StartPool2),
                 ?_assertEqual({ok, <<"value">>}, Set1),
                 ?_assertEqual({ok, <<"value2">>}, Set2),
                 ?_assertEqual({ok, <<"value">>}, Readback1),
                 ?_assertEqual({ok, <<"value2">>}, Readback2),
                 ?_assertEqual(ok, Flush1),
                 ?_assertEqual({ok, <<>>}, ReadbackFlushed1),
                 ?_assertEqual({ok, <<"value2">>}, ReadbackFlushed2),
                 ?_assertEqual([testpool, testpool2], GetPools),
                 ?_assertEqual(ok, StopPool1),
                 ?_assertEqual(ok, StopPool2)]
        end}.

pools_cleanup_at_app_stop_test_() ->
    {setup, fun setup/0, fun (_) -> kill_memcached_instances() end,
     fun ([F]) ->
                ?_test(begin
                        ok=memcache:start_pool(testpool, "localhost", 3333, 10, 10, true),
                        ok=memcache:start_pool(testpool2, "localhost", 3334, 10, 10, true),
                        ok=F(),
                        ?assertEqual(
                            [],
                            os:cmd("ps aux | grep memcached | grep localhost"
                                   " | grep 3333 | awk '{print $2}'")
                        ),
                        ?assertEqual(
                            [],
                            os:cmd("ps aux | grep memcached | grep localhost"
                                   " | grep 3334 | awk '{print $2}'")
                        )
                    end)
        end}.

pools_get_restarted_by_sup_test_() ->
    {setup, fun setup/0, fun cleanup/1, fun (_) ->
                ?_test(begin
                        ok=memcache:start_pool(testpool, "localhost", 3333, 10, 10, true),
                        kill_memcached_instances(),
                        % get call to a killed memcached to trigger the child restart and wait a bit
                        % for it to happen
                        ?assertMatch({error, {poolboy_error, _}}, memcache:get(testpool, any)),
                        ?assertEqual({ok, <<>>}, retry_get(testpool, any, 200, 10))
                    end)
        end}.

with_started_memcached_test_() ->
    {setup, fun setup/0, fun cleanup/1, fun (_) ->
                ?_test(begin
                        ok=memcache_pools_sup:start_memcache("localhost", 3333, []),
                        ok=memcache:start_pool(testpool, "localhost", 3333, 10, 10, false),
                        ?assertEqual({ok, <<>>}, memcache:get(testpool, any))
                    end)
        end}.

port_verification_test_() ->
    {setup, fun setup/0, fun cleanup/1, fun (_) ->
                ?_test(begin
                        ok=memcache:start_pool(testpool, "localhost", 3333, 10, 10, true),
                        ResOk=memcache_pools_sup:wait_for_memcache("localhost", 3333, 1, 100),
                        ResError=memcache_pools_sup:wait_for_memcache("localhost", 3334, 3, 100),
                        ?assertEqual(ok, ResOk),
                        ?assertMatch({error, {unable_to_connect_to_memcached, _, _}},
                                     ResError)
                    end)
        end}.

single_pool_restart_test_() ->
    {setup, fun setup/0, fun cleanup/1, fun (_) ->
                {timeout, 30, [?_test(begin
                        lists:foreach(fun (_) ->
                                    ?assertMatch(ok, memcache:start_pool(testpool, "localhost",
                                                                         3333, [{memory, 4}], 10, 10, true)),
                                    ?assertMatch(ok, memcache:stop_pool(testpool))
                            end, lists:seq(1, 20))
                    end)]}
        end}.

env_test_() ->
    {setup,
        fun () ->
                application:load(memcache),
                {ok, F1} = elibs_application:set_env(memcache, pools,
                    [{yaddapool, [
                                {size, 10},
                                {max_overflow, 20},
                                {port, 3333},
                                {host, "localhost"},
                                {opts, [{memory, 4}]},
                                {start_server, true}
                            ]
                        }
                    ]),
                [F1] ++ setup()
        end,
        fun cleanup/1,
        [
            ?_assertEqual(
                ok,
                memcache_pools_sup:wait_for_memcache("localhost", 3333, 1, 100)
            ),
            ?_assertEqual({ok, <<"cash">>}, memcache:set(yaddapool, age, <<"cash">>)),
            ?_assertEqual({ok, <<"cash">>}, memcache:get(yaddapool, age))
        ]
    }.


%%================================================================================================
%% Internal functions
%%================================================================================================

setup() ->
    error_logger:tty(false),
    {ok, F} = memcache:start(),
    [F].

cleanup(L) ->
    lists:foreach(fun(F) -> F() end, L),
    kill_memcached_instances().

kill_memcached_instances() ->
    memcache_pools_sup:stop_memcache("localhost", 3333),
    memcache_pools_sup:stop_memcache("localhost", 3334).

retry_get(Pool, Key, _Interval, 0) ->
    memcache:get(Pool, Key);
retry_get(Pool, Key, Interval, Retries) ->
    case memcache:get(Pool, Key) of
        {ok, _}=Ok ->
            Ok;
        {error, {poolboy_error, _}} ->
            ?debugFmt("Retrying get(~p, ~p). ~p retries left with ~p ms interval",
                      [Pool, Key, Retries - 1, Interval]),
            timer:sleep(Interval),
            retry_get(Pool, Key, Interval, Retries - 1)
    end.



src/memcache.erl
%% @doc
%% This module acts as entry point for the application.
-module(memcache).

-behaviour(gen_server).

-include("memcache.hrl").
-include_lib("erlanglibs/include/logging.hrl").

-define(DEFAULT_EXPIRATION_TIME, 3600).

%% API
-export([delete/2,
         flush/1,
         get/2,
         get_pools/0,
         stats/1,
         remove_all_pools/0,
         set/3,
         set/4,
         start/0,
         start_link/0,
         start_pool/6,
         start_pool/7,
         stop/0,
         stop_pool/1
        ]).

-type memcached_opts()::[{memory, non_neg_integer()}].
-type pool_name()::atom().
-type pool_host()::string().
-type pool_port()::pos_integer().
-type pool_size()::pos_integer().
-type pool_max_overflow()::pos_integer().
-type pool()::{pool_name(), pool_host(), pool_port(), pool_size(), pool_max_overflow(), boolean()}.
-type key()::binary()|term().
-type value()::binary()|term().
-type expiration()::pos_integer().
-export_type([pool_name/0, pool_host/0, pool_port/0, pool_size/0, pool_max_overflow/0, key/0,
              value/0, expiration/0]).

%% gen_server callbacks
-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).

-record(state,  {pools::ets:tab()}). %  The ets entries are instances of pool()

%%%===================================================================
%%% API
%%%===================================================================

%% @doc
%% Starts the application and all the necessary dependencies. It returns a function that stops
%% all that is needed in order to leave the environment as it was.
-spec start() -> {ok, fun (() -> ok)}.
%% @end
start() ->
    elibs_application:start(?MODULE).

%% @doc
%% Stops this application
-spec stop() -> ok.
%% @end
stop() ->
    application:stop(?MODULE).

%% @doc Starts the server
-spec start_link() -> {ok, pid()}.
%% @end
start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

%% @doc Starts a cache pool.
%% The StartServer variable decides whether this application takes care of starting the memcached
%% server as well. If it was already started, just returns ok. Note that the start server
%% functionality will only work for starting local memcached instances, so Host should be a
%% valid name for localhost or an IP address for one of the local network interfaces.
-spec start_pool(pool_name(), pool_host(), pool_port(), memcached_opts(),
                 pool_size(), pool_max_overflow(), boolean()) -> ok | {error, term()}.
%% @end
start_pool(Poolname, Host, Port, MemcachedOpts, Size, MaxOverflow, StartServer) ->
    gen_server:call(?MODULE, {start_pool, {Poolname, Host, Port, MemcachedOpts, Size, MaxOverflow, StartServer}}).

start_pool(Poolname, Host, Port, Size, MaxOverflow, StartServer) ->
    start_pool(Poolname, Host, Port, [], Size, MaxOverflow, StartServer).

%% @doc
%% Stops the given pool. If the memcached server for the pool had been started by this application,
%% it is stopped as well.
%% @end
-spec stop_pool(pool_name()) -> ok | {error, pool_not_found | term()}.
%%
stop_pool(Poolname) ->
    gen_server:call(?MODULE, {stop_pool, Poolname}).

%% @doc
%% retrieve the value with a certain binary key from the specified pool.
%% @end
-spec get(pool_name(), key()) -> {ok, value()} | {error, term()}.
get(Poolname, Key) ->
    Op = fun() ->
            Worker = poolboy:checkout(Poolname),
            case gen_server:call(Worker, {get, Key}) of
                {error, R} ->
                    {error, {poolboy_error, R}};
                Reply ->
                    poolboy:checkin(Poolname, Worker),
                    {ok, Reply}
            end
    end,
    run_in_pool(Poolname, Op).

%%
%% @doc
%% Stores the given value under the given key in the specified pool
%% Uses the default expiration time.
%% @end
-spec set(pool_name(), key(), binary() ) -> {ok, value()} | {error, term()}.
set(Poolname, Key, Value) ->
    set(Poolname, Key, Value, ?DEFAULT_EXPIRATION_TIME).

%% @doc
%% Stores the given value under the given key in the specified pool
%% After the given expiration time (in s), the entry is deleted from the pool.
%% @end
-spec set(pool_name(), key(), value(), expiration() ) -> value().
set(Poolname, Key, Value, Expiration) ->
    Op = fun () ->
            Worker = poolboy:checkout(Poolname),
            case gen_server:call(Worker, {set, Key, Value, Expiration}) of
                {error, R} ->
                    {error, {poolboy_error, R}};
                <<>> ->
                    poolboy:checkin(Poolname, Worker),
                    {ok, Value}
            end
    end,
    run_in_pool(Poolname, Op).

%% @doc
%% Deletes the data associated with a certain key in the specified pool. This functions returns ok
%% even if the key couldn't be found
%% @end
-spec delete(pool_name(), key()) -> ok | {error, term()}.
delete(Poolname, Key) ->
    Op = fun() ->
            Worker = poolboy:checkout(Poolname),
            case gen_server:call(Worker, {delete, Key}) of
                {error, R} ->
                    {error, {poolboy_error, R}};
                _ ->
                    poolboy:checkin(Poolname, Worker),
                    ok
            end
    end,
    run_in_pool(Poolname, Op).

%% @doc
%% Clears all cached entries in the given pool
%% @end
-spec flush(pool_name()) -> ok | {error, term()}.
flush(Poolname) ->
    Op = fun() ->
            Worker = poolboy:checkout(Poolname),
            case gen_server:call(Worker, flush) of
                {error, R} ->
                    {error, {poolboy_error, R}};
                _ ->
                    poolboy:checkin(Poolname, Worker),
                    ok
            end
    end,
    run_in_pool(Poolname, Op).


get_pools() ->
    gen_server:call(?MODULE, get_pools).

-type stats() :: [{atom(), string()}, ...].
%% @doc Pool statistics
-spec stats(pool_name()) -> stats() | {error, term()}.
%% @end
stats(Poolname) ->
    Op = fun() ->
            Worker = poolboy:checkout(Poolname),
            case gen_server:call(Worker, stats) of
                {error, R} ->
                    {error, {poolboy_error, R}};
                Ret ->
                    poolboy:checkin(Poolname, Worker),
                    Ret
            end
    end,
    run_in_pool(Poolname, Op).

%% @doc
%% Removes all the pools started via start_pool/7. This function is automatically called when this
%% application is about to stop.
-spec remove_all_pools() -> [{pool_name(), StopRes::ok | {error, term()}}].
%% @end
remove_all_pools() ->
    gen_server:call(?MODULE, remove_all_pools).

%%%===================================================================
%%% gen_server callbacks
%%%===================================================================

%% @private
-spec init([]) -> {ok, #state{}}.
init([]) ->
    ?MEMCACHE_POOLS_ETS = ets:new(?MEMCACHE_POOLS_ETS,
        [public, {read_concurrency, true}, named_table]),
    {ok, #state{pools=?MEMCACHE_POOLS_ETS}}.

%% @private
-type call_type()::{start_pool, pool()} | {stop_pool, pool_name()} |
    remove_all_pools | get_pools | term().
-spec handle_call(call_type(), any(), #state{}) ->
    {reply, ok | {error, term()}, #state{}} | {noreply, #state{}}.
handle_call({start_pool, {Poolname, Host, Port, MemcachedOpts, Size, MaxOverflow, StartServer}}, _From, State) ->
    Res = case check_pool_availability(Poolname) of
        ok ->
            case memcache_pools_sup:add_pool(Poolname, Host, Port, MemcachedOpts,
                                             Size, MaxOverflow, StartServer) of
                {ok, _Pid} ->
                    ok;
                {error, _}=E ->
                    E
            end;
        {error, _}=EE -> EE
    end,
    {reply, Res, State};
handle_call({stop_pool, Poolname}, _From, State) ->
    Res = case ets:lookup(?MEMCACHE_POOLS_ETS, Poolname) of
        [] ->
            {error, pool_not_found};
        [{Poolname, Host, Port, _Size, _MaxOverflow, StopServer}] ->
            case memcache_pools_sup:remove_pool(Poolname, Host, Port, StopServer) of
                ok ->
                    true = ets:delete(?MEMCACHE_POOLS_ETS, Poolname),
                    ok;
                {error, _}=E ->
                    E
            end
    end,
    {reply, Res, State};
handle_call(get_pools, _From, #state{pools=T}=State) ->
    Pools = ets:foldl(fun(X, Acc) -> [element(1, X)|Acc] end, [], T),
    {reply, Pools, State};
handle_call(remove_all_pools, _From, State) ->
    Res = lists:foldl(fun ({Poolname, Host, Port, _Size, _MaxOverflow, StopServer}, Acc) ->
                    PoolRes=memcache_pools_sup:remove_pool(Poolname, Host, Port, StopServer),
                    true = ets:delete(?MEMCACHE_POOLS_ETS, Poolname),
                    [{Poolname, PoolRes}| Acc]
            end, [], ets:tab2list(?MEMCACHE_POOLS_ETS)),
    {reply, Res, State};
handle_call(_Request, _From, State) ->
    {noreply, State}.

%% @private
-spec handle_cast(any(), #state{}) -> {noreply, #state{}}.
handle_cast(_Msg, State) ->
    {noreply, State}.

%% @private
-spec handle_info(any(), #state{}) -> {noreply, #state{}}.
handle_info(_Info, State) ->
    {noreply, State}.

%% @private
-spec terminate(any(), #state{}) -> ok.
terminate(_Reason, _State) ->
    ok.

%% @private
-spec code_change(any(), #state{}, any()) -> {ok, #state{}}.
code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

%%%===================================================================
%%% Internal functions
%%%===================================================================

-spec run_in_pool(pool_name(), fun (() -> term())) -> term() | {error, pool_not_found | term()}.
run_in_pool(Poolname, Op) ->
    case ets:lookup(?MEMCACHE_POOLS_ETS, Poolname) of
        [] -> {error, pool_not_found};
        [_] ->
            try
                Op()
            catch
                _:Reason ->
                    {error, {poolboy_error, Reason}}
            end
    end.

-spec check_pool_availability(pool_name()) ->
    ok | {error, poolname_in_use}.
check_pool_availability(Poolname) ->
    case ets:lookup(?MEMCACHE_POOLS_ETS, Poolname) of
        [] -> ok;
        [_] -> {error, poolname_in_use}
    end.





src/memcache_sup.erl
%% @doc
%% The main application supervisor. It manages both the memcache gen_server and the pools
%% supervisor with a one for all strategy.
-module(memcache_sup).

-behaviour(supervisor).

%% API
-export([start_link/0]).

%% Supervisor callbacks
-export([init/1]).

-define(CHILD(Id, Mod, Type, Args), {Id, {Mod, start_link, Args},
                                     permanent, 5000, Type, [Mod]}).

%%%===================================================================
%%% API functions
%%%===================================================================

-spec start_link() -> {ok, pid()}.
start_link() ->
    supervisor:start_link({local, ?MODULE}, ?MODULE, []).

%%%===================================================================
%%% Supervisor callbacks
%%%===================================================================

init([]) ->
    Children=[
        ?CHILD(memcache, memcache, worker, []),
        ?CHILD(memcache_pools_sup, memcache_pools_sup, supervisor, [])
    ],
    {ok, {{one_for_all, 5, 10}, Children}}.

%%%===================================================================
%%% Internal functions
%%%===================================================================


src/memcache_pools_sup.erl
%% @doc
%% Supervises the pools and memcached instances managed by the application, restarting them in case
%% of failure with a one for one strategy
-module(memcache_pools_sup).

-behaviour(supervisor).

-include("memcache.hrl").
-include_lib("erlanglibs/include/logging.hrl").

%% API
-export([add_pool/6,
         add_pool/7,
         remove_pool/4,
         start_link/0,
         start_pool/3
]).

%% Supervisor callbacks
-export([init/1]).

%% just for testing purposes
-ifdef('TEST').
-export([start_memcache/3, stop_memcache/2, wait_for_memcache/4]).
-endif.

-define(CHECK_PORT_RETRIES, 5).
-define(PORT_RETRY_INTERVAL, 200).
-define(CHILD(Id, Mod, Type, Args), {Id, {Mod, start_link, Args},
                                     permanent, 5000, Type, [Mod]}).

%%%===================================================================
%%% API functions
%%%===================================================================

%% @doc
%% Starts this supervisor. To be called from memcache_sup
-spec start_link() -> {ok, pid()}.
%% @end
start_link() ->
    supervisor:start_link({local, ?MODULE}, ?MODULE, []).


%% @doc
%% Adds a pool to the list of children of this supervisor
-spec add_pool(memcache:pool_name(), memcache:pool_host(), memcache:pool_port(),
               memcache:pool_size(), memcache:pool_max_overflow(), StartServer::boolean()) ->
    {ok, pid()} | {error, term()}.
%% @end
add_pool(Poolname, Host, Port, Size, MaxOverflow, StartServer) ->
    add_pool(Poolname, Host, Port, [], Size, MaxOverflow, StartServer).

%% @doc
%% Adds a pool to the list of children of this supervisor
-spec add_pool(memcache:pool_name(), memcache:pool_host(), memcache:pool_port(), memcache:memcached_opts(),
               memcache:pool_size(), memcache:pool_max_overflow(), StartServer::boolean()) ->
    {ok, pid()} | {error, term()}.
%% @end
add_pool(Poolname, Host, Port, MemcachedOpts, Size, MaxOverflow, StartServer) ->
    ChildSpec = build_child_spec(Poolname, Host, Port, MemcachedOpts, Size, MaxOverflow, StartServer),
    supervisor:start_child(?MODULE, ChildSpec).

%% @doc
%% Terminates a pool and removes its childspec from this supervisor
-spec remove_pool(memcache:pool_name(), memcache:pool_host(), memcache:pool_port(),
                  ServerManaged::boolean()) -> ok | {error, term()}.
%% @end
remove_pool(Poolname, Host, Port, true) ->
    ok=stop_memcache(Host, Port),
    remove_pool(Poolname, Host, Port, false);
remove_pool(Poolname, _Host, _Port, false) ->
    poolboy:stop(Poolname),
    case supervisor:terminate_child(?MODULE, Poolname) of
        ok ->
            ok = supervisor:delete_child(?MODULE, Poolname);
        {error, not_found} ->
            {error, not_found}
    end.

%% @private
%% @doc
%% This functions wraps the poolboy call to start the pool of connections with the start_memcache/3
%% call in order to start the server before if needed.
%%
%% If all is fine, registers the pool to ?MEMCACHE_POOL_ETS
-spec start_pool({start | already_started, memcache:pool_host(), memcache:pool_port(), memcache:memcached_opts()}
    | undefined, [{atom(), term()}], tuple()) -> {ok, pid()} | {error, term()}.
%% @end
start_pool({start, Host, Port, MemcachedOpts}, PoolboyOpts, CreateOpts) when is_list(Host), is_integer(Port) ->
    ?INFO("Starting memcache pool and memcached server at ~p", [{Host, Port}]),
    start_memcache(Host, Port, MemcachedOpts),
    case do_start_pool(Host, Port, PoolboyOpts) of
        {ok, Pid} ->
            true = ets:insert(?MEMCACHE_POOLS_ETS, [CreateOpts]),
            {ok, Pid};
        {error, _} = Err ->
            Err
    end;
start_pool({already_started, Host, Port, _}, PoolboyOpts, CreateOpts) when is_list(Host), is_integer(Port) ->
    ?INFO("Starting memcache pool", []),
    case do_start_pool(Host, Port, PoolboyOpts) of
        {ok, Pid} ->
            true = ets:insert(?MEMCACHE_POOLS_ETS, [CreateOpts]),
            {ok, Pid};
        {error, _} = Err ->
            Err
    end.

%%%===================================================================
%%% Supervisor Callbacks
%%%===================================================================

%% @private
init([]) ->
    Children = lists:map(
        fun ({Poolname, Props}) ->
                Size = proplists:get_value(size, Props),
                MaxOverflow = proplists:get_value(max_overflow, Props),
                Port = proplists:get_value(port, Props),
                Host = proplists:get_value(host, Props),
                MemcachedOpts = proplists:get_value(opts, Props, []),
                Start = proplists:get_value(start_server, Props),
                build_child_spec(Poolname, Host, Port, MemcachedOpts, Size, MaxOverflow, Start)
        end,
        get_config_value(pools, [])
    ),
    {ok, {{one_for_one, 5, 10}, Children}}.

%%%===================================================================
%%% Internal functions
%%%===================================================================

-spec build_child_spec(memcache:pool_name(), memcache:pool_host(), memcache:pool_port(),
                        memcache:pool_config(), memcache:pool_size(), memcache:pool_max_overflow(), boolean()) ->
    supervisor:child_spec().
build_child_spec(Poolname, Host, Port, MemcachedOpts, Size, MaxOverflow, StartServer) ->
    StartServerOpts=case StartServer of
        true -> {start, Host, Port, MemcachedOpts};
        false -> {already_started, Host, Port, MemcachedOpts}
    end,
    CreateOpts = {Poolname, Host, Port, Size, MaxOverflow, StartServer},
    PoolboyOpts = [{name, {local, Poolname}}, {worker_module, erlmc_conn}, {size, Size},
                   {max_overflow, MaxOverflow}, {host, Host}, {port, Port}],
    {Poolname, {?MODULE, start_pool, [StartServerOpts, PoolboyOpts, CreateOpts]},
     transient, 5000, worker, [poolboy]}.

-spec do_start_pool(memcache:pool_host(), memcache:pool_port(), [{atom(), term()}]) ->
    {ok, pid()} | {error, term()}.
do_start_pool(Host, Port, PoolboyOpts) ->
    case wait_for_memcache(Host, Port, ?CHECK_PORT_RETRIES, ?PORT_RETRY_INTERVAL) of
        ok -> start_poolboy(PoolboyOpts);
        {error, _}=Error -> Error
    end.

-spec start_poolboy([{atom(), term()}]) -> {ok, pid()} | {error, term()}.
start_poolboy(PoolboyOpts) ->
    try
        poolboy:start_link(PoolboyOpts)
    catch
        E:EE ->
            ?WARNING("Unable to start pool: ~p", [{E, EE}]),
            {error, EE}
    end.

-spec wait_for_memcache(memcache:pool_host(), memcache:pool_port(),
                        non_neg_integer(), pos_integer()) -> ok | {error, term()}.
wait_for_memcache(Host, Port, 0, _) ->
    {error, {unable_to_connect_to_memcached, Host, Port}};
wait_for_memcache(Host, Port, Retries, RetryInterval) ->
    case gen_tcp:connect(Host, Port, [binary, {reuseaddr, true}]) of
        {ok, Sock} ->
            gen_tcp:close(Sock),
            ok;
        {error, econnrefused} ->
            timer:sleep(RetryInterval),
            wait_for_memcache(Host, Port, Retries - 1, RetryInterval);
        {error, Reason} ->
            {error, {unable_to_connect_to_memcached, Reason}}
    end.

-spec start_memcache(memcache:pool_host(), memcache:pool_port(), memcache:pool_config()) -> ok | {error, term()}.
start_memcache(Host, Port, MemcachedOpts) ->
    Memory = proplists:get_value(memory, MemcachedOpts, ?MEMCACHE_DEFAULT_MEMORY_SIZE),
    Cmd=elibs_string:format("memcached -d -l ~s -m ~p -p ~p", [Host, Memory, Port]),
    %io:format("Starting : ~s\n", [Cmd]),
    case os:cmd(Cmd) of
        [] ->
            timer:sleep(100),
            ok;
        Other ->
            ?WARNING("Unexpected return starting memcached ~p: ~p", [{Host, Port}, Other]),
            {error, Other}
    end.

-spec stop_memcache(memcache:pool_host(), memcache:pool_port()) -> ok | {error, term()}.
stop_memcache(Host, Port) ->
    Cmd = elibs_string:format(
            "kill -9 $(ps aux | grep memcached | grep ~s | grep ~p | awk '{print $2}')",
            [Host, Port]
        ),
    case os:cmd(Cmd) of
        [] ->
            timer:sleep(100),
            ok;
        Other ->
            ?WARNING("Unexpected return stoping memcached ~p: ~p", [{Host, Port}, Other]),
            {error, Other}
    end.

get_config_value(Key, Default) ->
    case application:get_env(memcache, Key) of
        undefined -> Default;
        {ok, Val} -> Val
    end.